#### Senior Analyst assignment 
#### Data pre-processing

In [30]:



# Import packages

import pandas as pd
import sqlite3
from sqlalchemy.types import *



In [31]:


#Read dim_customer file
dim_customer = pd.read_csv("/Users/ragwendraacharya/Downloads/Senior Data Analyst Assignment - DimCustomer.csv",index_col=False)
dim_customer_2 = pd.read_csv("/Users/ragwendraacharya/Downloads/Senior Data Analyst Assignment - DimCustomer.csv",index_col=False)

#Read dim_product file
dim_product = pd.read_csv("/Users/ragwendraacharya/Downloads/Senior Data Analyst Assignment - DimProduct.csv",index_col=False)

#Read dimsalesterritory file
dimsales_ = pd.read_csv("/Users/ragwendraacharya/Downloads/Senior Data Analyst Assignment - DimSalesTerritory.csv",index_col=False)

#Read factressellerssales
factrsellers = pd.read_csv("/Users/ragwendraacharya/Downloads/Senior Data Analyst Assignment - FactResellerSales.csv",index_col=False)

#Read diminvoices file
dim_invoices = pd.read_csv("/Users/ragwendraacharya/Downloads/Senior Data Analyst Assignment - DimInvoices.csv",index_col=False)


#### Dim customer data set

In [32]:
#print the first 5 rows of dim_customer
print(dim_customer.head())

   CustomerKey  GeographyKey CustomerAlternateKey Title  FirstName MiddleName  \
0        11000            26           AW00011000   NaN        Jon          V   
1        11001            37           AW00011001   NaN     Eugene          L   
2        11002            31           AW00011002   NaN      Ruben        NaN   
3        11003            11           AW00011003   NaN    Christy        NaN   
4        11004            19           AW00011004   NaN  Elizabeth        NaN   

  LastName  NameStyle   BirthDate MaritalStatus  ... EnglishOccupation  \
0     Yang          0  11/11/1996             M  ...      Professional   
1    Huang          0  24/04/2022             S  ...      Professional   
2   Torres          0    9/4/1990             M  ...      Professional   
3      Zhu          0    2/5/1988             S  ...      Professional   
4  Johnson          0  29/03/1979             S  ...      Professional   

  SpanishOccupation FrenchOccupation  HouseOwnerFlag  NumberCarsOwne

In [33]:
#Print the file info 
print(dim_customer.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18484 entries, 0 to 18483
Data columns (total 29 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   CustomerKey           18484 non-null  int64 
 1   GeographyKey          18484 non-null  int64 
 2   CustomerAlternateKey  18484 non-null  object
 3   Title                 101 non-null    object
 4   FirstName             18484 non-null  object
 5   MiddleName            10654 non-null  object
 6   LastName              18484 non-null  object
 7   NameStyle             18484 non-null  int64 
 8   BirthDate             18484 non-null  object
 9   MaritalStatus         18484 non-null  object
 10  Suffix                3 non-null      object
 11  Gender                18484 non-null  object
 12  EmailAddress          18484 non-null  object
 13  YearlyIncome          18484 non-null  int64 
 14  TotalChildren         18484 non-null  int64 
 15  NumberChildrenAtHome  18484 non-null

We can observe that there are 18484 with 29 columns with some columns having NULL values, lets check which columns have null values

In [34]:
# check for Null values in each column
print(dim_customer.isnull().any())

CustomerKey             False
GeographyKey            False
CustomerAlternateKey    False
Title                    True
FirstName               False
MiddleName               True
LastName                False
NameStyle               False
BirthDate               False
MaritalStatus           False
Suffix                   True
Gender                  False
EmailAddress            False
YearlyIncome            False
TotalChildren           False
NumberChildrenAtHome    False
EnglishEducation        False
SpanishEducation        False
FrenchEducation         False
EnglishOccupation       False
SpanishOccupation       False
FrenchOccupation        False
HouseOwnerFlag          False
NumberCarsOwned         False
AddressLine1            False
AddressLine2             True
Phone                   False
DateFirstPurchase       False
CommuteDistance         False
dtype: bool


In [35]:
null_count = dim_customer.isnull().sum().sum()
print('Number of null values:', null_count)

Number of null values: 62866


There are quite a large number of null values in this dataset, lets replace them with non-null so that they do not cause issues while querying.
Since all of the above null value columns are String object types we will perform a common operation to replace them at once.

In [36]:
# Replace null values with empty string  : 
      
dim_customer.fillna(" ", inplace = True)

In [37]:
# Now lets check again for null values : 
null_count_2 = dim_customer.isnull().sum().sum()
print('Number of null values:', null_count_2)


Number of null values: 0


Now lets check for duplicates within our dataset, the only duplicates we are concerned about is with regards to the *primary key*. The customer dataset has a CustomerKey which can be uniquely used to identify a cusomter. Let's check if this key duplicated anywhere within the dataset.

In [38]:
#Check for CustomerKey duplicates
duplicate_rows = dim_customer.duplicated(subset=['CustomerKey'])
duplicate_rows.sum()


0

In [ ]:
In a similar we are going to perform pre-processing of other datasets before they are loaded into the database

#### Dim Product Dataset

In [39]:
#print the first 5 rows of dim_product
print(dim_product.head())

   ProductKey ProductAlternateKey  ProductSubcategoryKey  \
0           1             AR-5381                    NaN   
1           2             BA-8327                    NaN   
2           3             BE-2349                    NaN   
3           4             BE-2908                    NaN   
4           5             BL-2036                    NaN   

  WeightUnitMeasureCode SizeUnitMeasureCode            ProductName  \
0                   NaN                 NaN        Adjustable Race   
1                   NaN                 NaN           Bearing Ball   
2                   NaN                 NaN        BB Ball Bearing   
3                   NaN                 NaN  Headset Ball Bearings   
4                   NaN                 NaN                  Blade   

   StandardCost  FinishedGoodsFlag Color  SafetyStockLevel  ...  Weight  \
0           NaN                  0   NaN              1000  ...     NaN   
1           NaN                  0   NaN              1000  ...     

In [40]:
#Print the file info 
print(dim_product.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ProductKey             569 non-null    int64  
 1   ProductAlternateKey    569 non-null    object 
 2   ProductSubcategoryKey  360 non-null    float64
 3   WeightUnitMeasureCode  245 non-null    object 
 4   SizeUnitMeasureCode    217 non-null    object 
 5   ProductName            569 non-null    object 
 6   StandardCost           358 non-null    float64
 7   FinishedGoodsFlag      569 non-null    int64  
 8   Color                  315 non-null    object 
 9   SafetyStockLevel       569 non-null    int64  
 10  ReorderPoint           569 non-null    int64  
 11  ListPrice              358 non-null    float64
 12  Size                   263 non-null    object 
 13  SizeRange              263 non-null    object 
 14  Weight                 245 non-null    float64
 15  DaysTo

In [41]:
# check for Null values in each column
print(dim_product.isnull().any())

ProductKey               False
ProductAlternateKey      False
ProductSubcategoryKey     True
WeightUnitMeasureCode     True
SizeUnitMeasureCode       True
ProductName              False
StandardCost              True
FinishedGoodsFlag        False
Color                     True
SafetyStockLevel         False
ReorderPoint             False
ListPrice                 True
Size                      True
SizeRange                 True
Weight                    True
DaysToManufacture        False
ProductLine               True
DealerPrice               True
Class                     True
Style                     True
ModelName                 True
StartDate                 True
EndDate                   True
Status                    True
dtype: bool


We can observe that there are 569 entries with 24 columns with 17 columns having NULL values.
Since these columns are a mix of numeric and string we will have to handle them seperately.

In [42]:
#For columns with string objects type we convert NULLs to string values 'Unknown'

dim_product[["WeightUnitMeasureCode","SizeUnitMeasureCode","Color","Size","SizeRange","Class","Style","ModelName","Status"]] = dim_product[["WeightUnitMeasureCode","SizeUnitMeasureCode","Color","Size","SizeRange","Class","Style","ModelName","Status"]].fillna(value = "Unknown")


In [43]:
#For columns with string objects type we convert NULLs to numeric values '0.0'
dim_product[["DealerPrice","ProductLine", "Weight", "ListPrice", "StandardCost","ProductSubcategoryKey"]] = dim_product[["DealerPrice","ProductLine", "Weight", "ListPrice", "StandardCost","ProductSubcategoryKey"]].fillna(value = 0.0)

In [44]:
#For columns with date values we convert NULLs to empty string ' '

dim_product[["EndDate","StartDate"]] = dim_product[["EndDate","StartDate"]].fillna(value = " ")

In [45]:
# check for Null values in each column
print(dim_product.isnull().any())

ProductKey               False
ProductAlternateKey      False
ProductSubcategoryKey    False
WeightUnitMeasureCode    False
SizeUnitMeasureCode      False
ProductName              False
StandardCost             False
FinishedGoodsFlag        False
Color                    False
SafetyStockLevel         False
ReorderPoint             False
ListPrice                False
Size                     False
SizeRange                False
Weight                   False
DaysToManufacture        False
ProductLine              False
DealerPrice              False
Class                    False
Style                    False
ModelName                False
StartDate                False
EndDate                  False
Status                   False
dtype: bool


In [ ]:
Now lets check for duplicates for Product Key which unoquely identifies each product. Let's check if this key duplicated anywhere within the dataset.

In [46]:
#Check for duplicates

duplicate_rows = dim_product.duplicated(subset=['ProductKey'])
duplicate_rows.sum()

0

#### FactResellersSales Dataset

In [47]:
#print the first 5 rows of dim_product
print(factrsellers.head())

   ProductKey  OrderDateKey  DueDateKey  ShipDateKey  ResellerKey  \
0         349      20101229    20110110     20110105          676   
1         350      20101229    20110110     20110105          676   
2         351      20101229    20110110     20110105          676   
3         344      20101229    20110110     20110105          676   
4         345      20101229    20110110     20110105          676   

   EmployeeKey  PromotionKey  CurrencyKey  SalesTerritoryKey SalesOrderNumber  \
0          285             1          100                  5          SO43659   
1          285             1          100                  5          SO43659   
2          285             1          100                  5          SO43659   
3          285             1          100                  5          SO43659   
4          285             1          100                  5          SO43659   

   ...  ProductStandardCost  TotalProductCost  SalesAmount    TaxAmt  \
0  ...            1898.094

This dataset has 5 rows and 27 columns ,seems like dataset has no NULL values at first glance let's do a double check to confirm

In [48]:
# check for Null values in each column
print(factrsellers.isnull().any())

ProductKey               False
OrderDateKey             False
DueDateKey               False
ShipDateKey              False
ResellerKey              False
EmployeeKey              False
PromotionKey             False
CurrencyKey              False
SalesTerritoryKey        False
SalesOrderNumber         False
SalesOrderLineNumber     False
RevisionNumber           False
OrderQuantity            False
UnitPrice                False
ExtendedAmount           False
UnitPriceDiscountPct     False
DiscountAmount           False
ProductStandardCost      False
TotalProductCost         False
SalesAmount              False
TaxAmt                   False
Freight                  False
CarrierTrackingNumber    False
CustomerPONumber         False
OrderDate                False
DueDate                  False
ShipDate                 False
dtype: bool


In [49]:
#Check for duplicates
duplicate_rows = factrsellers.duplicated(subset=['SalesOrderNumber'])
duplicate_rows.sum()

57059

Looks like the *SalesOrderNumber* is duplicated indicating that it can be used toiniquely identify each row. In this case we will have to create **composite primary key** , which is a combination of 2 or more columns to uniquely identify a entry.

In this case we use a combination  *SalesOrderNumber* and *SalesOrderLineNumber* to create this composite primary key. Let's name this new primary key as *SalesOrderId*

In [50]:
# Apply the custom function to create a new column 'SalesOrderId'

def combine_columns(row):
    return str(row['SalesOrderLineNumber']) + row['SalesOrderNumber']


factrsellers['SalesOrderId'] = factrsellers.apply(combine_columns, axis=1)

In [51]:
factrsellers['SalesOrderId'].head()

0    1SO43659
1    2SO43659
2    3SO43659
3    4SO43659
4    5SO43659
Name: SalesOrderId, dtype: object

In [52]:
duplicate_rows = factrsellers.duplicated(subset=['SalesOrderId'])
duplicate_rows.sum()

0

In [ ]:
Using the above approach we achieved a unique composite primary key for the FactResellerSales dataset

In [53]:
#check the dataset for first 5 rows

dim_invoices.head()


invoice_id invoice_date product_type  \
0  d8dd452d-7b5d-4a69-9d25-9c55c0b908e6  Jan 1, 2022          NaN   
1  74af949a-ac7a-4bce-8fa6-e0410952ffda  Jan 2, 2022  EOR Monthly   
2  de489739-6bf6-4494-9cff-21f0ff5f6d2f  Jan 2, 2022  EOR Monthly   
3  608ca796-b1f3-4515-b2e8-404ffacaca58  Jan 3, 2022  EOR Monthly   
4  08ec570b-a42d-437e-990b-aedbea16de9d  Jan 4, 2022  EOR Monthly   

  is_recurring invoice_status discount  discounted_amount_usd  amount_usd  
0           No       invoiced      NaN                 4076.0        4076  
1          Yes        pending      20%                  173.6         217  
2           No       invoiced      NaN                  263.0         263  
3           No       invoiced      NaN                  300.0         300  
4           No       invoiced      NaN                  232.0         232

In [54]:
dim_invoices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   invoice_id             500 non-null    object 
 1   invoice_date           500 non-null    object 
 2   product_type           463 non-null    object 
 3   is_recurring           500 non-null    object 
 4   invoice_status         500 non-null    object 
 5   discount               36 non-null     object 
 6   discounted_amount_usd  500 non-null    float64
 7   amount_usd             500 non-null    int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 31.4+ KB


In [55]:
# check for Null values in each column
print(dim_invoices.isnull().any())

invoice_id               False
invoice_date             False
product_type              True
is_recurring             False
invoice_status           False
discount                  True
discounted_amount_usd    False
amount_usd               False
dtype: bool


Dataset has 500 entries with 8 colunms with 2 columns with NULL values, lets convert the NULL to non-null values

In [56]:
# For the product type we are going to replace NULL with 'Unknown' string value and for the discount column replace NULL with 0%

dim_invoices[["product_type"]] = dim_invoices[["product_type"]].fillna(value = "Unknown")
dim_invoices[["discount"]] = dim_invoices[["discount"]].fillna(value = "0%")


In [57]:
#Check for duplicates on 'Invoice_id' 

duplicate_rows = dim_invoices.duplicated(subset=['invoice_id'])
duplicate_rows.sum()

0

#### SalesTerritory Dataset

In [58]:
#Check the first 5 rows 

dimsales_.head()

SalesTerritoryKey  SalesTerritoryAlternateKey SalesTerritoryRegion  \
0                  1                           1            Northwest   
1                  2                           2            Northeast   
2                  3                           3              Central   
3                  4                           4            Southwest   
4                  5                           5            Southeast   

  SalesTerritoryCountry SalesTerritoryGroup  
0         United States       North America  
1         United States       North America  
2         United States       North America  
3         United States       North America  
4         United States       North America

In [59]:
#Check for duplicates on SalesTerritoryKey

print(dimsales_.isnull().any())

SalesTerritoryKey             False
SalesTerritoryAlternateKey    False
SalesTerritoryRegion           True
SalesTerritoryCountry          True
SalesTerritoryGroup            True
dtype: bool


In [60]:
#check the dataset info
dimsales_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 5 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   SalesTerritoryKey           11 non-null     int64 
 1   SalesTerritoryAlternateKey  11 non-null     int64 
 2   SalesTerritoryRegion        10 non-null     object
 3   SalesTerritoryCountry       10 non-null     object
 4   SalesTerritoryGroup         10 non-null     object
dtypes: int64(2), object(3)
memory usage: 572.0+ bytes


Since dataset has only 11 entries  and 5 columns , lets check last 5 rows for the NULL values

In [61]:
#Check the last 5 rows 

dimsales_.tail()

SalesTerritoryKey  SalesTerritoryAlternateKey SalesTerritoryRegion  \
6                   7                           7               France   
7                   8                           8              Germany   
8                   9                           9            Australia   
9                  10                          10       United Kingdom   
10                 11                           0                  NaN   

   SalesTerritoryCountry SalesTerritoryGroup  
6                 France              Europe  
7                Germany              Europe  
8              Australia             Pacific  
9         United Kingdom              Europe  
10                   NaN                 NaN

In [62]:
#Replace NULL values with 'Unknown' string 

dimsales_[["SalesTerritoryRegion","SalesTerritoryCountry","SalesTerritoryGroup"]] = dimsales_[["SalesTerritoryRegion","SalesTerritoryCountry","SalesTerritoryGroup"]].fillna(value = "Unknown")

In [63]:
print(dimsales_.isnull().any())

SalesTerritoryKey             False
SalesTerritoryAlternateKey    False
SalesTerritoryRegion          False
SalesTerritoryCountry         False
SalesTerritoryGroup           False
dtype: bool


In [64]:
#Check for duplicates on 'SalesTerritoryKey 

duplicate_rows = dimsales_.duplicated(subset=['SalesTerritoryKey'])
duplicate_rows.sum()

0

Now that we have datasets cleaned and checked , lets now write the records in the dataframe into SQL database.

In [65]:

#create a new database connection called remotedb_test.db
conn= sqlite3.connect("remote_test.db")


In [66]:
# Write the dim_customer dataframe into SQL table
dim_customer.to_sql(
                'dim_customer',             # Name of the sql table
                conn,                 # sqlite.Connection or sqlalchemy.engine.Engine
                if_exists='replace',
                dtype={'CustomerKey': 'INTEGER PRIMARY KEY'})

18484

In [67]:
# Write the dim_product dataframe into SQL table
dim_product.to_sql(
                'dim_product',             # Name of the sql table
                conn,                      # sqlite.Connection or sqlalchemy.engine.Engine
                if_exists='replace',
                dtype={'ProductKey': 'INTEGER PRIMARY KEY'})

569

In [68]:
# Write the dim_sales dataframe into SQL table
dimsales_.to_sql(
                'dim_sales',             # Name of the sql table
                conn,                      # sqlite.Connection or sqlalchemy.engine.Engine
                if_exists='replace',
                dtype={'SalesTerritoryKey': 'INTEGER PRIMARY KEY'})



11

In [69]:
# Write the dim_sales dataframe into SQL table
dim_invoices.to_sql(
                'dim_invoices',             # Name of the sql table
                conn,                      # sqlite.Connection or sqlalchemy.engine.Engine
                if_exists='replace',
                dtype={'Invoice_id': 'INTEGER PRIMARY KEY'})

500

In [71]:
factrsellers.to_sql(
                'fact_resellers',             # Name of the sql table
                conn,                      # sqlite.Connection or sqlalchemy.engine.Engine
                if_exists='replace',
                dtype={'SalesOrderId': 'TEXT PRIMARY KEY'})


60855